src/control/lib/spdk/src/nvme_control.c

```diff
@@ -135,79 +109,95 @@ wipe_ctrlr(struct ctrlr_entry *centry, struct ns_entry *nentry)
 	struct lba0_data	 data;
 	struct wipe_res_t	*res = NULL, *tmp = NULL;
 	int			 rc;
+	struct spdk_nvme_qpair	*qpair;
+	char			*buf;
 
-	while (nentry != NULL) {
-		res = init_wipe_res();
-		res->next = tmp;
-		tmp = res;
+	res = init_wipe_res();
 
-		res->ns_id = spdk_nvme_ns_get_id(nentry->ns);
+	/** convert pci addr to string */
+	rc = spdk_pci_addr_fmt(res->ctrlr_pci_addr, sizeof(res->ctrlr_pci_addr),
+			       &centry->pci_addr);
+	if (rc != 0) {
+		res->rc = -NVMEC_ERR_PCI_ADDR_FMT;
+		return res;
+	}
 
-		rc = spdk_pci_addr_fmt(res->ctrlr_pci_addr,
-				       sizeof(res->ctrlr_pci_addr),
-				       &centry->pci_addr);
-		if (rc != 0) {
-			res->rc = -NVMEC_ERR_PCI_ADDR_FMT;
-			return res;
-		}
+	/** allocate NVMe queue pair for the controller */
+	qpair = spdk_nvme_ctrlr_alloc_io_qpair(centry->ctrlr, NULL, 0);
+	if (qpair == NULL) {
+		snprintf(res->info, sizeof(res->info),
+			 "spdk_nvme_ctrlr_alloc_io_qpair()\n");
+		res->rc = -1;
+		return res;
+	}
 
-		nentry->qpair = spdk_nvme_ctrlr_alloc_io_qpair(centry->ctrlr,
-							       NULL, 0);
-		if (nentry->qpair == NULL) {
-			snprintf(res->info, sizeof(res->info),
-				 "spdk_nvme_ctrlr_alloc_io_qpair()\n");
-			res->rc = -1;
-			return res;
-		}
+	/** allocate a 4K page, with 4K alignment */
+	buf =  spdk_dma_zmalloc(4096, 4096, NULL);
+	if (buf == NULL) {
+		snprintf(res->info, sizeof(res->info),
+			 "spdk_dma_zmalloc()\n");
+		res->rc = -1;
+		spdk_nvme_ctrlr_free_io_qpair(qpair);
+		return res;
+	}
 
-		data.buf = spdk_zmalloc(0x1000, 0x1000, NULL,
-					SPDK_ENV_SOCKET_ID_ANY,
-					SPDK_MALLOC_DMA);
-		if (data.buf == NULL) {
-			snprintf(res->info, sizeof(res->info),
-				 "spdk_zmalloc()\n");
-			res->rc = -1;
-			spdk_nvme_ctrlr_free_io_qpair(nentry->qpair);
-			return res;
+	/** iterate over the namespaces and wipe them out individually */
+	while (nentry != NULL) {
+		uint32_t sector_size;
+
+		if (tmp == NULL) {
+			/** first iteration */
+			tmp = res;
+		} else {
+			/** allocate new res */
+			res = init_wipe_res();
+			res->next = tmp;
+			tmp = res;
 		}
-		data.write_result = LBA0_WRITE_PENDING;
+
+		/** retrieve namespace ID and sector size */
+		res->ns_id = spdk_nvme_ns_get_id(nentry->ns);
+		sector_size = spdk_nvme_ns_get_sector_size(nentry->ns);
+
+		data.result = LBA0_WRITE_PENDING;
 		data.ns_entry = nentry;
 
-		rc = spdk_nvme_ns_cmd_write(nentry->ns, nentry->qpair,
-					    data.buf, 0, /* LBA start */
-					    1, /* number of LBAs */
+		/** zero out the first 4K block */
+		rc = spdk_nvme_ns_cmd_write(nentry->ns, qpair,
```

Comment:
not sure about this change, we specifically want to zero just the first LBA block as that's what is required as per the spec. we don't need to erase the full 4k. it may not matter if we do in practice but I don't see this as a necessary change.

Reply:
I initially bumped it from 1 sector (512B or 4KB depending on ns) to 4K to verify whether this was the root cause of the issue. This wasn't. I eventually decided to keep it since i think it is very common for filesystem to store data in the first 4K (one cannot write < a page via a block interface) and i would rather be conservative, just in case.

Reply:
ok, sounds sensible